### PR TITLE
Add filtering by annotations on search

### PR DESCRIPTION
**Pull request by Itxaka**
Allows to add filtering by annotations by passing the -a flag and the
annotation in the format "KEY=VALUE"
    
So you can search like this:

`hypper search repo -a "catalog.cattle.io/certified=rancher"`

And it will only show charts that have that exact annotation.

Requires #113 

Signed-off-by: Itxaka <igarcia@suse.com>

### CODE REVIEW

File: cmd/hypper/search_test.go
Comment (viccuad) [inline cmd/hypper/search_test.go:79]: :rofl:  :rofl:  :rofl:  top!

File: docs/user/howto/searchannotations.md
Comment (viccuad) [inline docs/user/howto/searchannotations.md:31]: `witht he` -> `with the`

(doesn't allow me to do suggestions Oo)

File: pkg/search/search.go
Comment (viccuad) [inline pkg/search/search.go:90]: since on line 95 we call `o.applConstraint(data)` with the data that we generate here, I suppose it's an AND?
I trust the tests :man_shrugging: .

**Review by viccuad — DISMISSED** [review]
nice! looks nice and clean, the docs are slowly growing :D.

File: pkg/search/search.go
Comment (Itxaka) [reply pkg/search/search.go:90]: Goddamm, maybe this is not as explicit as I wanted to be. I know its an AND **currently**! But in the comment Im wondering about what we should do with this, should we treat as AND or maybe as OR? Should we only allow passing one search annotation? etc...

I will update the comment to make it clearer once the search PR is merged and I need to rebase this

**Comment by Itxaka** [comment]
Rebased!

File: pkg/search/search.go
Comment (viccuad) [inline pkg/search/search.go:17]:
```suggestion
/* Package search implements the search for charts in repos but extracts it into a package
```

Comment (Itxaka) [reply pkg/search/search.go:17]: will fix :clown_face:

File: pkg/search/search.go
Comment (dragonchaser) [inline pkg/search/search.go:78]: we should either resolve that TODO or remove it from the PR.

File: pkg/search/index.go
Comment (dragonchaser) [inline pkg/search/index.go:17]: why did you remove the package description?

**Review by dragonchaser — CHANGES_REQUESTED** [review]

File: pkg/search/index.go
Comment (Itxaka) [reply pkg/search/index.go:17]: Now it lives on the doc.go

When there is a bit of text around in the packages, I prefer to extract it to a doc.go file, seems nicer and all the description adn explanation are together.

File: pkg/search/search.go
Comment (Itxaka) [reply pkg/search/search.go:78]: still struggling with this, will probably test limiting to just one `-a` flag

**Review by mattfarina — CHANGES_REQUESTED** [review]
I was looking over #93 (the issue with the requirements for this). The idea is that Rancher has 1 charts repo and multiple versions of Rancher use it. Some chart versions will not work with some versions of Rancher. So, a version range of supported rancher versions on the chart could be used to filter for known supported versions.

To make this work we need to have more than a check on the existence of an annotation key/value pair.

For that same issue we need to expose this via a Go API. That can be a separate PR, though.

**Comment by Itxaka** [comment]
I dont understand @mattfarina 

This covers your user case correctly. You can use this to search for `-a "catalog.cattle.io/rancher-version=2.5.*" --regex ` for example and it would search for that with regex in versioning.

So this perfectly covers that.

Now, please, if you have a specific case in mind, **post it in the issue** because if the issue is not clear or not defined enough, we are wasting time writing code that has to be refactored.

This covers much more ground that the specific issue and its flexible enough to be used perfectly on hypper, plus brings value into hypper by making it all about annotations.

**Comment by viccuad** [comment]
What I understood from Matt is that, in addition to filtering on annotations, we want to have a semver-aware filter. So we could ask for things matching `~1.2.3`, or `^1.2.3`, etc.

**Comment by Itxaka** [comment]
> What I understood from Matt is that, in addition to filtering on annotations, we want to have a semver ranges aware filter. So we could ask for things matching ~1.2.3, or ^1.2.3, etc.

but then we are not being consequent here.

Either we want to use a **exact** key to search in the annotations and make that semver aware, or we want this to not be rancher agnostic.

Like if we are gonna search for `catalog.cattle.io/rancher-version` we are tying up ourselves to rancher, but its clear on the ticket that this is not desired....so...Im not sure.

If there is a key we want to search for that we know contains semver then we should say that in the ticket, otherwise we get a generic, not-tied-to-rancher, annotation filter search.

Sure, we can be reckless and just try to load the values from the annotations into a semver constraint and fail otherwise but that is such a brute force approach that it makes no sense to me.


Again, the ticket doesnt seem to be clear. Annotations can be of any type. If we want to search for just one annotation that contains the rancher version and do a semver constraint on there, good for me, I just need to know which annotation so we can discard the rest. Otherwise, lets build a more intelligent, generic annotation search that can be used to search also for **hypper** annotations!


We seem to have an identity issue in which we either want hypper to be this community tool with a clear goal and suddenly we move as making hypper a vessel for rancher use cases. And this makes the hypper vision not clear enough IMO.


In any case, I still think this bring value to hypper as it allows use to center ourselves on annotations and search for those.

File: pkg/search/search.go
Comment (dragonchaser) [reply pkg/search/search.go]: Can you add a card to the board for that?

Comment (Itxaka) [reply pkg/search/search.go]: This is outdated, doesnt exists anymore. Not sure why github doesnt automatically resolve this....

Comment (dragonchaser) [reply pkg/search/search.go]: cool, will resolve then :)

**Review by viccuad — APPROVED** [review]
Approving, the PR is nice and hast tests! If we need to change or add filters, we can do it in subsequent PRs.

**Review by dragonchaser — APPROVED** [review]

**Review by mattfarina — COMMENTED** [review]
Let's hold off on this until after the v0.1.0 release.

In our meeting with Darren he had wanted semver range resolution as a feature. It was more than exact key/value pair checking. I will message him to double check what they are looking for.

**Review by mattfarina — APPROVED** [review]

**Pull request merged**
